File: Day 022/Problem 652. Find Duplicate Subtrees.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    map<TreeNode*, string> mp;
    map<string, int> freq; 

    string serialize(TreeNode* root, vector<TreeNode*> &answer) {
        if (!root) return "#";

        string left = "";
        if (root->left && mp.find(root->left) != mp.end()) {
            left = mp[root->left];
        } else {
            left = serialize(root->left, answer);
        }

        string right = "";
        if (root->right && mp.find(root->right) != mp.end()) {
            right = mp[root->right];
        } else {
            right = serialize(root->right, answer);
        }

        string result = to_string(root->val) + "_" + left + "_" + right;
        mp[root] = result;
        freq[result]++;

        if (freq[result] == 2) {
            answer.push_back(root);
        }

        return result;
    }

    vector<TreeNode*> findDuplicateSubtrees(TreeNode* root) {
        vector<TreeNode*> result;
        serialize(root, result);

        return result;
    }
};

File: Day 093/Problem 652. Find Duplicate Subtrees.java
import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

class TreeNode {
    int val;
    TreeNode left;
    TreeNode right;
    TreeNode() {}
    TreeNode(int val) { this.val = val; }
    TreeNode(int val, TreeNode left, TreeNode right) {
        this.val = val;
        this.left = left;
        this.right = right;
    }
}

class Solution {
    
    Map<TreeNode, String> treeMap = new HashMap<>();
    List<TreeNode> result = new ArrayList<>();
    Map<String, Integer> seen = new HashMap<>();

    private String dfs(TreeNode root) {
        if (root == null) return "";
        if(treeMap.containsKey(root)) return treeMap.get(root);

        String str = new StringBuilder().append(root.val).append('#').append(dfs(root.left)).append('#').append(dfs(root.right)).toString();    
        if (seen.getOrDefault(str, 0) == 1) {
            result.add(root);
        }
        seen.put(str, seen.getOrDefault(str, 0) + 1);

        return str;
    }

    public List<TreeNode> findDuplicateSubtrees(TreeNode root) {
        dfs(root);
        return result;
    }
}